skule_vote/backend/test_ballot.py

```diff
@@ -0,0 +1,312 @@
+                "voter": voters[0],
+                "candidate": candidate,
+                "rank": 0,
+                "election": referendum,
+            },
+            {
+                "voter": voters[0],
+                "candidate": ron,
+                "rank": 1,
+                "election": referendum,
+            },
+            {  # Second vote (2 candidates ranked)
+                "voter": voters[1],
+                "candidate": ron,
+                "rank": 0,
+                "election": referendum,
+            },
+            {
+                "voter": voters[1],
+                "candidate": candidate,
+                "rank": 1,
+                "election": referendum,
+            },
+            {  # Third vote (1 candidate ranked)
+                "voter": voters[2],
+                "candidate": ron,
+                "rank": 0,
+                "election": referendum,
+            },
+            {  # Fourth vote (1 candidate ranked)
+                "voter": voters[3],
+                "candidate": candidate,
+                "rank": 0,
+                "election": referendum,
+            },
+            {  # Spoiled ballot
+                "voter": voters[4],
+                "candidate": None,
+                "rank": None,
+                "election": referendum,
+            },
+        ]
+
+        # Create and serialize the ballots
+        for ballot in ballots:
+            Ballot.objects.create(**ballot)
+        ballot_dict = BallotSerializer(Ballot.objects.all())
+
+        # Ballot dict gives us Candidate objects, we need indicies from the choices list.
+        ballots_formatted = ballot_dict.map_candidates_in_ballots_to_choices(
+            ballots=ballot_dict.data, choices=choices
+        )
+
+        # The choice array is a queryset, we need a dictionary
+        choices_dict = CandidateSerializer(choices, many=True).data
+        for i in range(len(choices_dict)):
+            choices_dict[i] = dict(choices_dict[i])
+
+        # Calculate the results and assert
+        results = calculate_results(
+            ballots=ballots_formatted,
+            choices=choices_dict,
+            numSeats=referendum.seats_available,
+        )
+        self.assertEqual(results["winners"], ["NO (TIE)"])
+        self.assertEqual(results["rounds"][0]["Lisa Li"], 2)
+        self.assertEqual(results["rounds"][0]["Reopen Nominations"], 2)
+        self.assertEqual(results["quota"], 3)
+        self.assertEqual(results["spoiledBallots"], 1)
+
+        # Total votes is len(voters) - NUM_SPOILED because
+        # 1. We consider rankings of multiple candidates for the *same* position
+        #   as a single vote (i.e., voters[0] ranking two candidates is 1 vote.
+        # 2. We don't consider spoiled ballots as votes
+        self.assertEqual(results["totalVotes"], len(voters) - NUM_SPOILED)
+
+    # YES/NO election with some ballots being errors
+    def test_one_candidate_ballot_errored(self):
```

Comment:
I wonder if this should error out instead of returning the error count? probably something to discuss with them?